extensions/amp-instagram/1.0/instagram.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {ContainWrapper} from '../../../src/preact/component';
+import {dict} from '../../../src/utils/object';
+import {getData} from '../../../src/event-helper';
+import {parseJson} from '../../../src/json';
+import {useLayoutEffect, useRef, useState} from '../../../src/preact';
+
+/**
+ * @param {!InstagramPropsDef} props
+ * @return {PreactDef.Renderable}
+ */
+export function Instagram({
+  shortcode,
+  captioned,
+  style,
+  title,
+  requestResize,
+  ...rest
+}) {
+  const iframeRef = useRef(null);
+  const [heightStyle, setHeightStyle] = useState(dict());
+  const [opacity, setOpacity] = useState(0);
+
+  useLayoutEffect(() => {
+    /**
+     * @param {Event} event
+     */
+    function handleMessage(event) {
+      if (
+        event.origin != 'https://www.instagram.com' ||
+        event.source != iframeRef.current.contentWindow
+      ) {
+        return;
+      }
+
+      const data = parseJson(getData(event));
+
+      if (data['type'] == 'MEASURE') {
+        const height = data['details']['height'];
+        if (requestResize) {
+          requestResize(height);
+        } else {
+          setHeightStyle(dict({'height': height}));
+        }
+        setOpacity(1);
```

Comment:
So I'm looking at the old implementation. And it also sets `opacity` on iframe loaded. Should we do it here as well? We can pass `onLoad` event to the `<iframe>` and do it there.

Reply:
We definitely can. I had set the `opacity` at this point to prevent CLS. That way the user doesn't see any shifting of the iframe (because height only changes once a message is received from Instagram) once it's visible. What are your thoughts on that? Want to confirm before I add.

Reply:
The concern that I had is whether this message arrives _always_. E.g. what if the initial height was exactly what the embed needed, would it send this message?

Reply:
Yup, the message always arrives. Even if the height is initially set to the correct height, a message is still received and processed. I tested this by manually setting the initial height to the correct height for the iframe+shortcode combo and seeing if a height is still received, and it is.

Reply:
Still relying on this feels like relying on the internal implementation details of this embed. The worst that can happen then is that an embed would never show vs it will show but maybe a slightly wrong size. If they had a clear "ready to show" message, that'd be a different story. I think we still need `onLoad` and we can set a timer to unblock opacity in case resize failed to do so.

Reply:
Here's a little comparison video I made. To the naked eye, it's really hard to tell a difference. However, when slowed down, it looks like setting opacity `onLoad` makes an adjustment after the iframe is visible, causing the actual picture to load a split second after the iframe shows up on the page.

Reply:
That's definitely expected. If we were to do this via iframe.onload, it'd not be either/or, but rather we'd do both and will delay opacity reset from onload a bit. I.e. we'd do something like "update opacity to 1 at the earliest of resize-message-received or onload+500ms". I.e. it's a [double tap](https://www.youtube.com/watch?v=HQe2f8L-Rkc).

But not to get overly paranoid, I'm ok if we skip this for now.